modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/IgniteH2Indexing.java

```diff
@@ -1104,7 +1150,31 @@ public void bindParameters(PreparedStatement stmt, @Nullable Collection<Object>
 
         twoStepQry.pageSize(qry.getPageSize());
 
-        QueryCursorImpl<List<?>> cursor = new QueryCursorImpl<>(queryTwoStep(cctx, twoStepQry, cctx.keepBinary()));
+        final AtomicReference<GridAbsClosure> mapQrysCancel = new AtomicReference<>();
+        final AtomicReference<GridAbsClosure> rdcQryCancel = new AtomicReference<>();
+
+        QueryCursorImpl<List<?>> cursor = new QueryCursorImpl<>(
+            doQueryTwoStep(cctx, twoStepQry, cctx.keepBinary(), qry.getTimeout(), mapQrysCancel, rdcQryCancel),
+            new GridAbsClosure() {
+                @Override public void apply() {
+                    GridAbsClosure clo;
+                        // Cancellation closure is changed on each attempt. Make sure we use correct closure.
+                        while(true) {
```

Comment:
This loop looks strange to me. Is cancellation closure really can be reset multiple times? Then probably you never know that you used the correct one.